balancer/pickfirstleaf/pickfirstleaf.go

```diff
@@ -0,0 +1,658 @@
+	// subconn.
+	if activeSD, found := b.subConns.Get(sd.addr); !found || activeSD != sd {
+		return
+	}
+	if newState.ConnectivityState == connectivity.Shutdown {
+		return
+	}
+
+	if newState.ConnectivityState == connectivity.Ready {
+		b.shutdownRemaining(sd)
+		if !b.addressList.seekTo(sd.addr) {
+			// This should not fail as we should have only one subconn after
+			// entering READY. The subconn should be present in the addressList.
+			b.logger.Errorf("Address %q not found address list in  %v", sd.addr, b.addressList.addresses)
+			return
+		}
+		b.state = connectivity.Ready
+		b.cc.UpdateState(balancer.State{
+			ConnectivityState: connectivity.Ready,
+			Picker:            &picker{result: balancer.PickResult{SubConn: sd.subConn}},
+		})
+		return
+	}
+
+	// If the LB policy is READY, and it receives a subchannel state change,
+	// it means that the READY subchannel has failed.
+	if b.state == connectivity.Ready && newState.ConnectivityState != connectivity.Ready {
+		// Once a transport fails, the balancer enters IDLE and starts from
+		// the first address when the picker is used.
+		b.state = connectivity.Idle
+		b.addressList.reset()
+		b.cc.UpdateState(balancer.State{
+			ConnectivityState: connectivity.Idle,
+			Picker:            &idlePicker{exitIdle: b.ExitIdle},
+		})
+		return
+	}
+
+	if b.firstPass {
+		switch newState.ConnectivityState {
+		case connectivity.Connecting:
+			// The balancer can be in either IDLE, CONNECTING or TRANSIENT_FAILURE.
+			// If it's in TRANSIENT_FAILURE, stay in TRANSIENT_FAILURE until
+			// it's READY. See A62.
+			// If the balancer is already in CONNECTING, no update is needed.
+			if b.state == connectivity.Idle {
+				b.state = connectivity.Connecting
+				b.cc.UpdateState(balancer.State{
+					ConnectivityState: connectivity.Connecting,
+					Picker:            &picker{err: balancer.ErrNoSubConnAvailable},
+				})
+			}
+		case connectivity.TransientFailure:
+			sd.lastErr = newState.ConnectionError
+			// Since we're re-using common subconns while handling resolver updates,
+			// we could receive an out of turn TRANSIENT_FAILURE from a pass
+			// over the previous address list. We ignore such updates.
+
+			if curAddr := b.addressList.currentAddress(); !equalAddressIgnoringBalAttributes(&curAddr, &sd.addr) {
+				return
+			}
+			if b.addressList.increment() {
+				b.requestConnection()
+				return
+			}
+			// End of the first pass.
+			b.endFirstPass(newState.ConnectionError)
+		case connectivity.Idle:
+			// A subconn can transition from CONNECTING directly to IDLE when
+			// a transport is successfully created, but the connection fails before
+			// the subconn can send the notification for READY. We treat this
+			// as a successful connection and transition to IDLE.
+			if curAddr := b.addressList.currentAddress(); !equalAddressIgnoringBalAttributes(&sd.addr, &curAddr) {
+				return
+			}
+			b.state = connectivity.Idle
+			b.addressList.reset()
```

Comment:
There are 2 conditions:
1. If the `curAddr` is in TF, `requestConnection` will move to the next address in the list.
2. If the curAddr is not in TF, it has to be in CONNECTING or IDLE. SHUTDOWN is not possible as closed subconns are removed from the map immediately and `requestConnection` is not called when a READY subconn is present. If `curAddr` is in IDLE, connect is called to transition it to CONNECTING.

Due to point 1, the previous state of `curAddr` can never be TF. 
Due to point 2, we can assume that the subconn  will always enter CONNECTING. Here, we handle all transitions out of CONNECTING.

Reply:
Oh, I see. Because this is only operating on `currentAddress` then we know it was either `READY` or `CONNECTING` when we got past the previous `if`?  Because `TF` would have already updated the `currentAddress`?

Reply:
Yes, that's right. The READY->IDLE case is handled even earlier in this function by the condition `if b.state == connectivity.Ready && newState.ConnectivityState != connectivity.Ready`.